src/syntax/ppx_sedlex.ml

```diff
@@ -253,24 +264,228 @@ let gen_recflag auto =
     Nonrecursive
   with Exit -> Recursive
 
+let gen_offsets traces i = function
+  | (_, []), _ -> None
+  | (_, aliases), _ ->
+      let n = List.length aliases in
+      let _, trans, finals = traces.(i) in
+      let cases =
+        List.map (fun ({ actions; _ } : Sedlex.trans_case) -> actions) trans
+        @ List.map (fun ({ actions; _ } : Sedlex.final_case) -> actions) finals
+      in
+      let action2cases = Hashtbl.create n in
+      List.iteri
+        (fun i actions ->
+          List.iter
+            (fun action ->
+              try
+                let offsets = Hashtbl.find action2cases action in
+                Hashtbl.replace action2cases action (i :: offsets)
+              with Not_found -> Hashtbl.add action2cases action [i])
+            actions)
+        cases;
+      let counter = ref 0 in
+      let alias2offset = Hashtbl.create n in
+      let cases2offset = Hashtbl.create 31 in
+      Hashtbl.iter
+        (fun action offsets ->
+          try
+            let i = Hashtbl.find cases2offset offsets in
+            Hashtbl.replace cases2offset offsets i;
+            Hashtbl.add alias2offset action i
+          with Not_found ->
+            Hashtbl.add cases2offset offsets !counter;
+            Hashtbl.add alias2offset action !counter;
+            incr counter)
+        action2cases;
+      Some (!counter, alias2offset)
+
+let gen_cset ~loc x char_set =
+  let interval a b =
+    [%expr
+      [%e eint ~loc a] <= [%e evar ~loc x]
+      && [%e evar ~loc x] <= [%e eint ~loc b]]
+  in
+  match char_set with
+    | (a, b) :: l ->
+        List.fold_left
+          (fun acc (a, b) -> [%expr [%e acc] || [%e interval a b]])
+          (interval a b) l
+    | [] -> assert false
+
+let gen_trace lexbuf traces i = function
+  | None -> []
+  | Some (offsets_num, action_offsets) ->
+      let loc = default_loc in
+      let initial, trans, finals = traces.(i) in
+      let offset_array =
+        value_binding ~loc
+          ~pat:[%pat? __sedlex_offsets]
+          ~expr:(pexp_array ~loc (List.init offsets_num (fun _ -> [%expr 0])))
```

Comment:
I would use an invalid position here (e.g. '-1`) as it would make it more obvious if we ever encounter a bug where position are not updated.

Reply:
This is done